Flaky Jetpack E2E: fix flaky interaction of the editor block toolbar for the AI Assistant block flow.

worldomonation opened the pull request:
Related to: https://github.com/Automattic/wp-calypso/issues/81547 and https://github.com/Automattic/wp-calypso/issues/81547.

Fixes https://github.com/Automattic/wp-calypso/issues/81547.

## Proposed Changes

This PR uses the `elementHandle` and `waitForElementState` to wait for the teditor toolbar element to be stable before interaction.

## Testing Instructions

Ensure the following build configurations are passing:
  - [x] Calypso E2E (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [x] Jetpack E2E Simple (desktop)
  - [ ] Jetpack E2E AT (desktop)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- add wait for element state

karenroldan commented on packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts:
```diff
@@ -100,11 +102,16 @@ export class AIAssistantFlow implements BlockFlow {
-		await block
-			.getByRole( 'button', { name: 'Stop request' } )
-			.waitFor( { state: 'detached', timeout: 30 * 1000 } );
+		const handle = await block.elementHandle();
```
**suggestion:** It looks like Playwright does not recommend using [elementHandle](https://playwright.dev/docs/api/class-elementhandle) anymore.  

We could maybe use `block.waitFor({ state: 'visible', timeout: TIMEOUT }),` instead:

```
await Promise.all( [
	block
		.getByRole( 'button', { name: 'Stop request' } )
		.waitFor( { state: 'detached', timeout: TIMEOUT } ),
	block.waitFor({ state: 'visible', timeout: TIMEOUT }),
] );
```

worldomonation replied:
@karenroldan Playwright does indeed advocate against using anything related to ElementHandle, and have for some time.

However, one thing I found is that `waitFor` does not have an equivalent to the `waitForElementState('stable')`. I don't know why this is the case and it's something I should raise with the maintainers. 

`waitFor( {state: 'visible'})` is not quite equivalent - it doesn't wait for the element to reach a state of being stable ([where the bounding box doesn't change for at least two consecutive animation frames](https://playwright.dev/docs/actionability#stable)). 

If we were using Playwright Test Runner, we could use auto-retrying assertions to wait for the editor toolbar to be stable.

worldomonation replied:
I'm going to see if there's another event we can key off that doesn't rely on using ElementHandle.

karenroldan replied:
@worldomonation Thanks for pointing that out! We also don't want to add a fixed delay to account for potential instability. 

We can continue using the `elementHandle` for now. It doesn't introduce any immediate issues or drawbacks. 👍🏻

worldomonation replied:
I'm trying out another way so will hold off on merging for now.

worldomonation replied:
Okay, so after a few more tries - it still looks like there's no replacement for `ElementHandle.waitForElementState('stable')`.

This is a quick stress test results using ElementHandle:
![image](https://github.com/Automattic/wp-calypso/assets/6549265/d4a8b206-a19e-4db3-912f-21d1a3c75edf)

This was the best I could do without using ElementHandle:
![image](https://github.com/Automattic/wp-calypso/assets/6549265/51bc0156-4c26-4848-a085-348a384f021e)

As you can see, there are always at least two failures in each stress test run without using ElementHandle.

worldomonation replied:
I plan to open an issue/question with Playwright maintainers on why they have not implemented equivalent to the `waitForElementState` for Locators.

karenroldan replied:
I appreciate the deep dive you took on this and thanks for sharing the stress test results! Keeping the tests stable is definitely a priority. 👏🏻

worldomonation committed:
packages/calypso-e2e/.eslintrc.js
- add exception to rule

packages/calypso-e2e/src/lib/components/editor-block-toolbar-component.ts
- add waitFor

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- add try/catch to stop the generation request once the timeout is met.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- re-extend timeout to 30s

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- different timeout for AT and Simple.
- waitFor a bunch of selectors

worldomonation committed:
packages/calypso-e2e/src/lib/components/editor-popover-menu-component.ts
- add waitFor before clicking on the popover item.

packages/calypso-e2e/src/lib/components/editor-block-toolbar-component.ts
- revert to using elementhandle.
